crates/ra_hir_def/src/body/lower.rs

```diff
@@ -84,6 +85,40 @@ struct ExprCollector<'a> {
     source_map: BodySourceMap,
 }
 
+macro_rules! track_parent {
+    (@build $collector:ident, $parent:expr $(,)?) => {
+        $parent
+    };
+    (@build $collector:ident, $parent:expr, opt $expr:ident $($rest:tt)*) => {
+        {
+            if let Some(expr) = $expr {
+                $collector.body.parent_map.insert(expr, $parent);
+            }
+            track_parent!(@build $collector, $parent $($rest)*)
+        }
+    };
+    (@build $collector:ident, $parent:expr, vec $expr:ident $($rest:tt)*) => {
+        {
+            for expr in $expr {
+                $collector.body.parent_map.insert(expr, $parent);
+            }
+            track_parent!(@build $collector, $parent $($rest)*)
+        }
+    };
+    (@build $collector:ident, $parent:expr, $expr:ident $($rest:tt)*) => {
+        {
+            $collector.body.parent_map.insert($expr, $parent);
+            track_parent!(@build $collector, $parent $($rest)*)
+        }
+    };
+    ($collector:ident, $parent:expr, $($rest:tt)*) => {
+        {
+            let parent = $parent;
+            track_parent!(@build $collector, parent, $($rest)*)
+        }
+    }
+}
```

Comment:
In general, we use macros sparingly in rust-analyzer. For "just remove some copy paste" there's almost always a better solution, which typically is:

* just write a bit of boilerplate
* write some type-level or runtime code to do this. 

In this case, I think what would work is some runtime tracking. For example, we could allocate a single `Vec` with children or something like that. i am not entirely sure about this, but probability that the macro is the only non-allocating solution seems pretty low to me. 

Additionally, in this case I think there's an obvious easy solution -- let's remove parent tracking altogether, and instead turn an `assert` about parents into just the code that computes the parents mapping.

Reply:
Thanks to your persuasion, I came up with something much smaller. It takes advantage of `walk_child_exprs` and the only downside is it walks all immediate child exprs of an expr even if it already found an answer (`walk_child_exprs` doesn't support early return).